### ISSUE TITLE
results

### ISSUE THREAD

**Issue by ajbalogh** [open-traffic-generator/models#14]
Stats results should go back to returning simple key-value pairs instead of returning table view – because it really adds more boilerplate to the script.
	• One has to find indices of each column name
	• One has to make an assumption on type of returned value and cast it accordingly (since everything is returned as strings)
 
I think it’s possible to implement GraphQL-like query-response in OpenAPI (to fetch only intended column).
The possible issues I can think of are:
	• Same key-names inside stat for each flow/port will be returned
	• We might by default expose all stats that are not applicable for a given use case (but this should not be much of a problem since users will focus on stats they’re interested in)
	 
e.g. see snippet below for comparison in minimum use-case
``` 
 # fetch stats
result = results_api.get_port_results(
    ResultPortRequest(
        port_names=[p.name for p in PORTS],
        columns=['name', 'frames_tx', 'frames_rx']
    )
)
 
# extract and compare stats using key-value pair
assert result.rows[0]['frames_tx'] == result.rows[1]['frames_rx'] == count
 
# extract and compare stats using table view
tx = result.columns.index('frames_tx')
rx = result.columns.index('frames_rx')
assert int(result.rows[0][tx]) == count and int(result.rows[1][rx]) == count
```

**Comment by ashutshkumr** [comment]
@ajbalogh, @winstonliu1111, @ankur-sheth I have compiled minimal number of port / flow stats we need to expose over API (I just copy-pasted names used internally, so please excuse them).

**Kindly review** if in our data models, we need to include some or get rid of some properties (and their types).
e.g. We definitely need to include latency measurements but I cannot confirm if first/last timestamp is required by user.

Another minute detail: `*Bytes` and `*ByteRate` are not L1 properties. i.e. for a 64B frame, TxBytes will be 64 and not 64+8+12.

```go
type PortStats struct {
	TxFrames          uint64
	TxFrameRate       uint64
	TxBytes           uint64
	TxByteRate        uint64
	TxLoss            float64
	RxFrames          uint64
	RxFrameRate       uint64
	RxBytes           uint64
	RxByteRate        uint64
	RxLoss            float64
	RxPFCClass0Frames uint64
	RxPFCClass1Frames uint64
	RxPFCClass2Frames uint64
	RxPFCClass3Frames uint64
	RxPFCClass4Frames uint64
	RxPFCClass5Frames uint64
	RxPFCClass6Frames uint64
	RxPFCClass7Frames uint64
	RxVLANFrames      uint64
	RxIPv4Packets     uint64
	RxTCPPackets      uint64
	RxUDPPackets      uint64
	FirstTimestampNS  uint64
	LastTimestampNS   uint64
}
```

```go
type PGIDStats struct {
	// traffic stats
	RxFrames         uint64
	RxFrameRate      uint64
	RxBytes          uint64
	RxByteRate       uint64
	RxLoss           float64
	RxMinLatency     uint64
	RxMaxLatency     uint64
	RxAvgLatency     uint64
	FirstTimestampNS uint64
	LastTimestampNS  uint64
	// traffic stats - TBD how to fetch
	TxFrames    uint64
	TxFrameRate uint64
	TxBytes     uint64
	TxByteRate  uint64
}
```

**Comment by ankur-sheth** [comment]
@ashutshkumr understood on `*Byte `and `*ByteRate` being L2 statistics.  I think that is correct and in line with existing implementations that don't count preamble and so on.

What stats are `TxLoss` and `RxLoss`?  Also we won't be supporting `RxPFCClass*` stats with Athena so I think it is ok to note them as not being supported in our capabilities API.

Ok on the `Tx*` stats in the PGID stats.  Btw, we shouldn't refer to that structure as PGID stats.  Call them Flow stats.

**Comment by ajbalogh** [comment]
Is it feasible to do txbytes on pgidstats when you have random/incrementing frame sizes?

**Comment by ashutshkumr** [comment]
> @ashutshkumr understood on `*Byte `and `*ByteRate` being L2 statistics. I think that is correct and in line with existing implementations that don't count preamble and so on.
> 
> What stats are `TxLoss` and `RxLoss`? Also we won't be supporting `RxPFCClass*` stats with Athena so I think it is ok to note them as not being supported in our capabilities API.
> 
> Ok on the `Tx*` stats in the PGID stats. Btw, we shouldn't refer to that structure as PGID stats. Call them Flow stats.

@ankur-sheth,
- Regarding `RxLoss`, we need a way to represent frames that have bad CRC (and are not counted as Received Frames). Should we expose `RxBadCRCFrames` instead ?
- Regarding `TxLoss`, on a second thought I don't think it makes much sense for Tx to report this. I had included it because DPDK exposes `tx_dropped`, which may not be supported by other traffic services.

**Comment by ashutshkumr** [comment]
> Is it feasible to do txbytes on pgidstats when you have random/incrementing frame sizes?

@ajbalogh, I agree with that concern too. I mentioned it because in data models it's exposed as `bytes_tx` for Flow Statistics. We can sort of interpolate Tx Frame count, but not sure how we'll determine exact tx bytes. I had checked IxN for reference, and it doesn't display Tx Bytes at all (but still shows Tx Byte rate). None of this information is readily available from traffic service.
